frontend_tests/puppeteer_lib/common.js

```diff
@@ -120,6 +125,167 @@ class CommonUtils {
+                      re-rendered based on server info?
+            */
+            const last_msg = current_msg_list.last();
+
+            if (last_msg.raw_content !== content) {
+                return false;
+            }
+
+            if (last_msg.locally_echoed) {
+                return false;
+            }
+
+            const row = rows.last_visible();
+
+            if (rows.id(row) !== last_msg.id) {
+                return false;
+            }
+
+            /*
+                Make sure the message is completely
+                re-rendered from its original "local echo"
+                version by looking for the star icon.  We
+                don't add the star icon until the server
+                responds.
+            */
+            return row.find('.star').length === 1;
+        }, content);
+    }
+
+    // Wait for any previous send to finish, then send a message.
+    async send_message(page, type, params) {
+        // If a message is outside the view, we will skip
+        // validation later.
+        const outside_view = params.outside_view;
+        delete params.outside_view;
+
+        await page.waitForSelector('#compose-send-button:enabled');
+        await page.waitForSelector('#compose-textarea');
+
+        if (type === "stream") {
+            await page.keyboard.press('KeyC');
+        } else if (type === "private") {
+            await page.keyboard.press("KeyX");
+            const recipients = params.recipient.split(', ');
+            for (let i = 0; i < recipients.length; i += 1) {
+                await this.set_pm_recipient(page, recipients[i]);
+            }
+            delete params.recipient;
+        } else {
+            assert.fail("`send_message` got invalid message type");
+        }
+
+        if (params.stream) {
+            params.stream_message_recipient_stream = params.stream;
+            delete params.stream;
+        }
+
+        if (params.subject) {
+            params.stream_message_recipient_topic = params.subject;
+            delete params.subject;
+        }
+
+        await this.fill_form(page, 'form[action^="/json/messages"]', params);
+        await this.turn_off_press_enter_to_send(page);
+        await page.waitForSelector("#compose-send-button", {visible: true});
+        await page.click('#compose-send-button');
+
+        // confirm if compose box is empty.
+        await page.waitForSelector('#compose-textarea');
+        const compose_box_element = await page.$("#compose-textarea");
+        const compose_box_content = await page.evaluate(element => element.textContent,
+                                                        compose_box_element);
+        assert.equal(compose_box_content, '', 'Compose box not empty after message sent');
+
+        if (!outside_view) {
+            await this.wait_for_message_fully_processed(page, params.content);
+        }
+
+        page.evaluate(() => {
+            compose_actions.cancel();
```

Comment:
Maybe add a comment saying that we're just closing the compose box here after our hopefully successful send.